src/frontend/src/optimizer/rule/apply_proj_filter.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::super::plan_node::*;
+use super::{BoxedRule, Rule};
+use crate::expr::{CorrelatedInputRef, Expr, ExprImpl, ExprRewriter, InputRef};
+use crate::optimizer::PlanRef;
+use crate::utils::Condition;
+
+/// This rule is for pattern: Apply->Project->Filter.
+/// To unnest, we just pull predicates contain correlated variables in Filter into Apply, and
+/// convert it into corresponding type of Join.
+pub struct ApplyProjFilterRule {}
+impl Rule for ApplyProjFilterRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let apply = plan.as_logical_apply()?;
+
+        let right = apply.right();
+        let project = right.as_logical_project()?;
+        let (mut exprs, mut expr_alias) = project.clone().decompose();
+        let begin = exprs.len();
+
+        let input = project.input();
+        let filter = input.as_logical_filter()?;
+
+        // Remove expressions contain correlated_input_ref from LogicalFilter.
+        let mut cor_exprs = vec![];
+        let mut uncor_exprs = vec![];
+        filter.predicate().clone().into_iter().for_each(|expr| {
```

Comment:
a `for` loop is more simple here :)

Reply:
[partition_map](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.partition_map) is another choice 🤪